File: src/lib/api/task.ts
import { call } from "./index";
import type { Task, CreateTaskRequest, UpdateTaskRequest } from "@/types";

export const taskApi = {
  getAll: (): Promise<Task[]> => call<Task[]>("get_tasks"),

  getById: (id: string): Promise<Task> => call<Task>("get_task", { id }),

  create: (task: CreateTaskRequest): Promise<Task> =>
    call<Task>("create_task", { task }),

  update: (id: string, task: UpdateTaskRequest): Promise<Task> =>
    call<Task>("update_task", { id, task }),

  delete: (id: string): Promise<void> => call<void>("delete_task", { id }),

  toggle: (id: string, enabled: boolean): Promise<Task> =>
    call<Task>("toggle_task", { id, enabled }),

  test: (id: string): Promise<string> => call<string>("test_task", { id }),
};

File: src-tauri/src/commands/task.rs
use std::sync::Arc;
use tauri::State;
use crate::database::{Database, dao::task::{Task, CreateTaskRequest, UpdateTaskRequest, TaskDao}};
use crate::error::Result;

#[tauri::command]
pub fn get_tasks(db: State<'_, Arc<Database>>) -> Result<Vec<Task>> {
    let conn = db.conn().lock().unwrap();
    TaskDao::get_all(&conn)
}

#[tauri::command]
pub fn get_task(db: State<'_, Arc<Database>>, id: String) -> Result<Task> {
    let conn = db.conn().lock().unwrap();
    TaskDao::get_by_id(&conn, &id)?
        .ok_or_else(|| crate::error::ToolsError::TaskNotFound(id))
}

#[tauri::command]
pub fn create_task(db: State<'_, Arc<Database>>, task: CreateTaskRequest) -> Result<Task> {
    let conn = db.conn().lock().unwrap();
    TaskDao::create(&conn, task)
}

#[tauri::command]
pub fn update_task(db: State<'_, Arc<Database>>, id: String, task: UpdateTaskRequest) -> Result<Task> {
    let conn = db.conn().lock().unwrap();
    TaskDao::update(&conn, &id, task)
}

#[tauri::command]
pub fn delete_task(db: State<'_, Arc<Database>>, id: String) -> Result<()> {
    let conn = db.conn().lock().unwrap();
    TaskDao::delete(&conn, &id)
}

#[tauri::command]
pub fn toggle_task(db: State<'_, Arc<Database>>, id: String, enabled: bool) -> Result<Task> {
    let conn = db.conn().lock().unwrap();
    TaskDao::toggle(&conn, &id, enabled)
}